File: backend/src/main/java/backend/entity/Project/Project.java
package backend.entity.Project;

import backend.dto.project.ProjectUpdateDTO;
import backend.entity.S3File;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Formula;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Getter
@Builder
@Entity
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class Project {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String projectName;
    private String description;

    @CreatedDate
    private LocalDateTime createdDate;

    // 삭제 관련
    private boolean isDeleted;
    private LocalDate deletedDate;
    private String deleteUsername;


    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL,  orphanRemoval = true)
    private List<ProjectMember> projectMembers =  new ArrayList<>();

    @Formula("(SELECT count(*) FROM ProjectMember pm WHERE pm.project_id = id)")
    private int memberCount;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL,  orphanRemoval = true)
    private List<ProjectInvite> projectInvites =  new ArrayList<>();

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL,  orphanRemoval = true)
    private List<S3File> s3Files =  new ArrayList<>();

    public void updateProjectInfo(ProjectUpdateDTO projectUpdateDTO){
        this.projectName = projectUpdateDTO.getProjectName();
        this.description = projectUpdateDTO.getDescription();
    }

    public void deleteProject(String username){
        this.isDeleted = true;
        this.deletedDate = LocalDate.now();
        this.deleteUsername = username;
    }

    public void restoreProject(){
        this.isDeleted = false;
        this.deletedDate = null;
        this.deleteUsername = null;
    }
}

File: frontend/src/types/project.js
/**
 * 프로젝트 내 멤버 요약 타입
 * @typedef {object} ProjectMemberSummary
 * @property {string} username   - 사용자 아이디
 * @property {string} nickname   - 사용자 닉네임
 * @property {'LEADER' | 'MEMBER'} role - 프로젝트 내 권한
 */

/**
 * 프로젝트 상세 데이터 모델
 * @see Project.java
 * @see ../../../../../backend/src/main/java/backend/dto/project/ProjectDTO.java
 * @typedef {object} ProjectData
 * @property {number} projectId                 - 프로젝트 고유 ID
 * @property {string} projectName               - 프로젝트 제목
 * @property {string} description               - 프로젝트 상세 설명
 * @property {number} taskProgress              - 프로젝트 전체 진행률 (0~100)
 * @property {number} memberCount               - 참여 중인 멤버 수
 * @property {string} creationDate               - 생성 일시 (ISO 문자열)
 * @property {string} [updatedDate]             - 최종 수정 일시 (ISO 문자열)
 * @property {boolean} isDeleted                - 삭제 여부
 * @property {string | null} [deletedDate]      - 삭제 일시 (ISO 문자열)
 * @property {string | null} deleteUsername     - 삭제 수행자 아이디
 * @property {ProjectMemberSummary[]} members   - 프로젝트 참여 멤버 목록
 */

/**
 * 프로젝트 목록 정보 DTO
 * @typedef {object} ProjectListDTO
 * @property {number} projectId
 * @property {string} projectName
 * @property {string} description
 * @property {number} memberCount
 * @property {number} taskProgress
 * @property {string} creationDate
 * @property {string} updatedDate
 */

export {};
